R/recursive_unit_tests.R

```diff
@@ -0,0 +1,156 @@
+RecursiveUnitTests <- R6::R6Class("RecursiveUnitTests",       # nolint
+  public = list(
+    initialize = function(path, filter = "test-.+\\.R$", recursive = TRUE) {
+      if (length(path) > 1) {
+        cli::cli_abort("Please provide a single path.")
+      }
+      private$path <- path
+      private$filter <- filter
+      private$recursive <- recursive
+
+      private$get_valid_test_paths()
+    },
+    run_tests = function(...) {
+      if (private$is_single_test_file()) {
+        testthat::test_file(path = private$valid_paths, ...)
+      } else if (private$is_single_test_dir()) {
+        testthat::test_dir(path = private$valid_paths, ...)
+      } else if (private$is_multiple_test_dirs()) {
+        private$test_recursive(...)
+      } else {
+        cli::cli_abort("Test run failed!")
+      }
+    }
+  ),
+  private = list(
+    filter = NULL,
+    path = NULL,
+    valid_paths = NULL,
+    recursive = TRUE,
+    get_valid_test_paths = function() {
+      if (fs::is_file(private$path)) {
+        private$valid_paths <- fs::path_filter(private$path, regexp = private$filter)
+      }
+
+      if (fs::is_dir(private$path)) {
+        valid_paths <- unique(
+          fs::path_dir(
+            fs::dir_ls(path = private$path,
+                       regexp = private$filter,
+                       recurse = private$recursive, type = "file")
+          )
+        )
+
+        private$valid_paths <- valid_paths[order(valid_paths)]
+      }
+
+      if (length(private$valid_paths) == 0) {
+        cli::cli_abort("No valid test file/s found in {.var {private$path}}.",
+                       call = rlang::caller_env(n = 3))
+      }
+    },
+    is_single_test_file = function() {
+      length(private$valid_paths) == 1 && fs::is_file(private$valid_paths)
+    },
+    is_single_test_dir = function() {
+      length(private$valid_paths) == 1
+    },
+    is_multiple_test_dirs = function() {
+      length(private$valid_paths) > 1
+    },
+    run_recursive_test_dir = function(...) {
+      t(
+        sapply(private$valid_paths, function(this_path) {
+          private$cat_cr()
+          cli::cat_line("Test Directory: ", this_path)
+
+          single_test_result <- as.data.frame(
+            testthat::test_dir(path = this_path, stop_on_failure = FALSE, ...))
+
+          colSums(single_test_result[, c("failed", "warning", "skipped", "passed")])
+        })
+      )
+    },
+    show_final_line = function(test_results) {
+      final_line_results <- colSums(test_results)
+
+      cli::cat_line(
+        summary_line(final_line_results[["failed"]],
+                             final_line_results[["warning"]],
+                             final_line_results[["skipped"]],
+                             final_line_results[["passed"]])
```

Comment:
Recommendation: Use the following indentation style when arguments don't fit on one line:
```r
func(
  long_arg_1, # All arguments indented with 2 spaces.
  long_arg_2
) # Closing parenthesis on new line, with the same indentation as `func`.
```

This way if `func` is renamed there is no need to adjust the indentation of its arguments, which is easy to forget about and leads to larger / not meaningful diffs (harder to review).